src/player.cpp

```diff
@@ -4678,15 +4678,22 @@ bool player::is_dead_state() const
 
 void player::on_dodge( Creature *source, int difficulty )
 {
-    if( difficulty == INT_MIN && source != nullptr ) {
-        difficulty = source->get_melee();
-    }
+    // Each avoided hit consumes an available dodge
+    // When no more available we are likely to fail player::dodge_roll
+    dodges_left--;
 
-    if( difficulty > 0 ) {
-        practice( skill_dodge, difficulty );
-    }
+    // Even if we are not to train still call practice to prevent skill rust
+    practice( skill_dodge, std::min( difficulty, 0 ) );
 
     ma_ondodge_effects();
+
+    // For adjacent attackers check for techniques usable upon successful dodge
+    if( source && g->m.valid_move( pos(), source->pos() ) ) {
```

Comment:
If you want to check for "adjacent attackers", why not use something like `square_dist(a, b) == 1` or similar? `map::valid_move` does way more than just checking for being adjacent, which somehow contradicts the comment.